### PR TITLE
Feature Avatar on Blog Posts

### PR DESCRIPTION
**Pull request by avgupt**
@zregvart I was not sure how to import avatar images, so I used a default image for now. Can you please help me with that?
Also, here's a screenshot of how it looks now.
image

### CODE REVIEW

**Comment by zregvart** [comment]
@avgupt the [issue](https://issues.apache.org/jira/browse/CAMEL-14675) mentions using avatar pictures from GitHub or from Gravatar.

File: antora-ui-camel/src/css/doc.css [hunk 1]
```diff
@@ -634,3 +634,15 @@ kbd,
...
+.avatar {
+  vertical-align: middle;
+  width: 100px;
+  height: 100px;
+  border-radius: 50%;
+}
+
+#publishDate {
+  position: relative;
+  bottom: 35px;
```
Comment (zregvart) [inline antora-ui-camel/src/css/doc.css]: Any changes to the design of the blog need to go to `blog.css`.

File: layouts/blog/post.html [hunk 2]
```diff
@@ -9,8 +9,15 @@ <h1><a href="{{.Page.Permalink}}">{{ .Title }}</a></h1>
...
-            <p>by <span rel="author">{{ .Params.author }}</span></p>
+            <table>
+                <tr>
+                    <td rowspan="2"> <img src="https://www.w3schools.com/howto/img_avatar.png" alt="Avatar" class="avatar"> </td>
+                    <td><p>by <span rel="author">{{ .Params.author }}</span></p></td>
+                </tr>
+                <tr id = "publishDate">
+                    <td><time itemprop="published" datetime="{{ dateFormat "2006-01-02" .PublishDate }}" title="{{ dateFormat "Monday, January 2, 2006" .PublishDate }}">{{ dateFormat "January 2, 2006" .PublishDate }}</time></td>
+                </tr>
+            </table>
```
Comment (zregvart) [inline layouts/blog/post.html]: We try to use semantic HTML, introducing tables for this is not the right approach.

**Comment by avgupt** [comment]
> @avgupt the [issue](https://issues.apache.org/jira/browse/CAMEL-14675) mentions using avatar pictures from GitHub or from Gravatar.

@zregvart Can you please help me in linking users to their GitHub images? Will it require changes in the camel repo?

**Comment by zregvart** [comment]
@avgupt look into GitHub developer documentation, there is an API to [fetch user details](https://developer.github.com/v3/users/#get-a-single-user) that includes the avatar image of the user. As the issue mentions this can be done using Hugo data templates. I think the change we need besides in the layouts and CSS is that for each blog post we need to add additional front matter attribute to hold the authors github username.

**Comment by avgupt** [comment]
@zregvart Thanks for the help! I have made some changes, adding front matter attribute `github_handle` to the blog posts would be enough to show the corresponding avatar. (Provided accessing GitHub is allowed by the website content security policy).
It's working perfectly on my local machine though.

File: antora-ui-camel/src/css/blog.css [hunk 3]
```diff
@@ -119,6 +119,24 @@ article.blog p {
...
+  position: relative;
+  bottom: 35px;
+}
+
+.date-and-name {
+  position: relative;
+  bottom: 105px;
+  left: 110px;
+}
+
```
Comment (zregvart) [inline antora-ui-camel/src/css/blog.css]: - These rules need to be scoped to `.blog`, as all other rules in this file are.
- Don't use `px`, we exclusively use relative lengths
- Don't use id-based rules for elements that could occur more than once

File: layouts/blog/post.html [hunk 4]
```diff
@@ -9,8 +9,24 @@ <h1><a href="{{.Page.Permalink}}">{{ .Title }}</a></h1>
...
+                        {{ end }}
+                    </div>
+                <div class="date-and-name">
+                    <p>by <span rel="author">{{ .Params.author }}</span></p>
+                    <time itemprop="published" datetime="{{ dateFormat "2006-01-02" .PublishDate }}" title="{{ dateFormat "Monday, January 2, 2006" .PublishDate }}">{{ dateFormat "January 2, 2006" .PublishDate }}</time>
+                </div>
+                {{ else }}
+                    <time itemprop="published" datetime="{{ dateFormat "2006-01-02" .PublishDate }}" title="{{ dateFormat "Monday, January 2, 2006" .PublishDate }}">{{ dateFormat "January 2, 2006" .PublishDate }}</time>
+                    <p>by <span rel="author">{{ .Params.author }}</span></p>
+                {{ end }}
```
Comment (zregvart) [inline layouts/blog/post.html]: These two blocks are almost identical, don't duplicate markup makes it more difficult to maintain. No need to reorder elements with modern CSS such as flex layout

File: layouts/blog/post.html [hunk 5]
```diff
@@ -9,8 +9,24 @@ <h1><a href="{{.Page.Permalink}}">{{ .Title }}</a></h1>
     <div class="post">
         <aside>
             <div class="summary">{{ .Params.preview }}</div>
-            <time itemprop="published" datetime="{{ dateFormat "2006-01-02" .PublishDate }}" title="{{ dateFormat "Monday, January 2, 2006" .PublishDate }}">{{ dateFormat "January 2, 2006" .PublishDate }}</time>
-            <p>by <span rel="author">{{ .Params.author }}</span></p>
+            <div class="info">
+                {{ if .Params.github_handle }}
+                    <div class="post-avatar">
+                        {{ $urlPre := "https://api.github.com" }}
+                        {{ $gist := getJSON $urlPre "/users/" .Params.github_handle }}
```
Comment (zregvart) [inline layouts/blog/post.html]: No need for `$urlPre` variable. Use better and more descriptive variable names than `$urlPre` and `$gist`.

File: layouts/blog/post.html [hunk 6]
```diff
@@ -9,8 +9,24 @@ <h1><a href="{{.Page.Permalink}}">{{ .Title }}</a></h1>
...
+                        {{ with $gist }}
```
Comment (zregvart) [inline layouts/blog/post.html]: Not sure there's a need for the `with` block here, it encompasses only one element

**Comment by avgupt** [comment]
@zregvart Thanks for the review! I have made the suggested changes. Please have a look.

File: antora-ui-camel/src/css/blog.css (hunk 3 above)
Comment (avgupt) [reply antora-ui-camel/src/css/blog.css]: Replaced `px` with `em`.
Removed id-based rules and scoped the rules to .blog

**Comment by avgupt** [comment]
@zregvart Kind reminder. :)

**Comment by avgupt** [comment]
@zregvart @oscerd Any suggestions?

File: layouts/blog/post.html [hunk 7]
```diff
@@ -9,8 +9,22 @@ <h1><a href="{{.Page.Permalink}}">{{ .Title }}</a></h1>
...
+                {{ $avatar := newScratch }}
+                {{ $avatar.Set "url" "_/img/avatar.svg" }}
+
+                {{ if .Params.github_user }}
+                    {{ $userData := getJSON "https://api.github.com/users/" .Params.github_user }}
+                    {{ $avatar.Set "url" $userData.avatar_url }}
+                {{ end }}
+                    <div class="post-avatar">
+                            <img src="{{ $avatar.Get "url" | relURL }}" class="avatar">
+                    </div>
```
Comment (zregvart) [inline layouts/blog/post.html]: I would do this without the scratch and without as many divs, like (haven't tested):

```suggestion
                {{ if .Params.github_user }}
                    {{ $userData := getJSON "https://api.github.com/users/" .Params.github_user }}
                    <img src="{{ $userData.avatar_url }}" class="avatar" alt="{{ .Params.author }}">
                {{ else }}
                    <img src="{{ "/_/img/avatar.svg" | relURL }} " class="avatar" alt="{{ .Params.author }}">
                {{ end }}
```

File: antora-ui-camel/src/img/avatar.svg [hunk 8]
```diff
@@ -0,0 +1,89 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
```
[the remark this answers is not in the document]
Comment (avgupt) [reply antora-ui-camel/src/img/avatar.svg]: I took the image from here: https://en.m.wikipedia.org/wiki/File:Missing_avatar.svg
Can I go ahead and optimze it?

File: layouts/blog/post.html [hunk 9]
```diff
@@ -9,8 +9,18 @@ <h1><a href="{{.Page.Permalink}}">{{ .Title }}</a></h1>
...
-            <p>by <span rel="author">{{ .Params.author }}</span></p>
+            <div class="info">
+                {{ if .Params.github_user }}
+                    {{ $userData := getJSON "https://api.github.com/users/" .Params.github_user }}
+                    <img src="{{ $userData.avatar_url }}" class="avatar" alt="{{ .Params.author }}">
+                {{ else }}
+                    <img src="{{ "/_/img/avatar.svg" | relURL }} " class="avatar" alt="{{ .Params.author }}">
+                {{ end }}
+                <div class="date-and-name">
+                    <p>by <span rel="author">{{ .Params.author }}</span></p>
```
Comment (zregvart) [inline layouts/blog/post.html]: This bit changed in #292 to allow multiple authors on a blog post, can we make the same change here and in the `github_user` parameter?

**Comment by zregvart** [comment]
This was used to further develop this feature in ac458a0f1f2a5f28e62a208f7460c7e757e1a0a1 84bcfe5d870930f25d5a0945026628a4541f642e a4c4cb74b7d7388fdcf0f9ce7b1faca67aead2f2 4f021ae6b6ff3a2b7dc100edad8357ac8802823c and e3671bfab39a7142a81c5e9d3e5fe6945b8f6200

**Pull request closed without merging**